include/particle.tcc

```diff
@@ -161,6 +192,18 @@ mpm::HDF5Particle mpm::Particle<Tdim>::hdf5() const {
 
   particle_data.cell_id = this->cell_id();
 
+  particle_data.material_id = this->material_id();
+
+  // Write state variables
+  if (material_ != nullptr) {
+    particle_data.nstate_vars = state_variables_.size();
+    unsigned i = 0;
```

Comment:
I think it's good to throw an error in case `state_variables_.size()` return more than 20, cause your particle_data cant handle that.